cedar-lean/Cedar/Thm/Validation/Validator.lean

```diff
@@ -0,0 +1,401 @@
+  evaluate (Expr.getAttr e attr) request entities
+:= by
+  simp only [EvaluatesSubst] at ih₁
+  simp only [substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁]
+
+theorem evaluates_subst_hasAttr {e : Expr} {attr : Attr} {request : Request} {entities : Entities}
+(ih₁ : EvaluatesSubst e request entities) :
+  evaluate (substituteAction request.action (Expr.hasAttr e attr)) request entities =
+  evaluate (Expr.hasAttr e attr) request entities
+:= by
+  simp only [EvaluatesSubst] at ih₁
+  simp only [substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁]
+
+theorem evaluates_subst_unaryApp {op : UnaryOp} {e : Expr} {request : Request} {entities : Entities}
+(ih₁ : EvaluatesSubst e request entities) :
+  evaluate (substituteAction request.action (Expr.unaryApp op e)) request entities =
+  evaluate (Expr.unaryApp op e) request entities
+:= by
+  simp only [EvaluatesSubst] at ih₁
+  simp only [substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁]
+
+theorem evaluates_subst_binaryApp {op : BinaryOp} {a b : Expr} {request : Request} {entities : Entities}
+(ih₁ : EvaluatesSubst a request entities)
+(ih₂ : EvaluatesSubst b request entities) :
+  evaluate (substituteAction request.action (Expr.binaryApp op a b)) request entities =
+  evaluate (Expr.binaryApp op a b) request entities
+:= by
+  simp only [EvaluatesSubst] at *
+  simp only [substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁, ih₂]
+
+theorem evaluates_subst_and {a b : Expr} {request : Request} {entities : Entities}
+(ih₁ : EvaluatesSubst a request entities)
+(ih₂ : EvaluatesSubst b request entities) :
+  evaluate (substituteAction request.action (Expr.and a b)) request entities =
+  evaluate (Expr.and a b) request entities
+:= by
+  simp only [EvaluatesSubst, substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁, ih₂]
+
+theorem evaluates_subst_or {a b : Expr} {request : Request} {entities : Entities}
+(ih₁ : EvaluatesSubst a request entities)
+(ih₂ : EvaluatesSubst b request entities) :
+  evaluate (substituteAction request.action (Expr.or a b)) request entities =
+  evaluate (Expr.or a b) request entities
+:= by
+  simp only [EvaluatesSubst, substituteAction, mapOnVars] at *
+  simp only [evaluate, ih₁, ih₂]
+
+theorem substitute_action_nil_set : ∀ (uid : EntityUID),
+  substituteAction uid (.set []) = .set []
+:= by
+  intro uid
+  simp only [substituteAction, mapOnVars, List.attach_def, List.pmap, List.map_nil]
+
+theorem substitute_action_cons_set : ∀ (h : Expr) (t : List Expr) (uid : EntityUID),
+  substituteAction uid (.set (h :: t)) = .set ((substituteAction uid h) :: t.map (substituteAction uid))
+:= by
+  intro h t uid
+  simp only [substituteAction, mapOnVars, List.attach_def, List.map_pmap_subtype, List.map_cons]
+  simp only [Expr.set.injEq, List.cons.injEq, true_and]
+  unfold substituteAction
+  simp only
+
+theorem forall_map {xs : List Expr} {f g : Expr → Result Value}
```

Comment:
We have a theorem in `List.lemmas` that is a generalized version of this.  See `List.mapM_congr`.  I suggest using the existing one and deleting this one.

Reply:
omg ... perfect